README.md
# repose cookbook

[![Build Status](https://travis-ci.org/rackerlabs/cookbook-repose.svg?branch=master)](https://travis-ci.org/rackerlabs/cookbook-repose)

This cookbook installs, configures and manages Repose.

"REPOSE is an open-source RESTful HTTP proxy service that scales to the cloud. REPOSE provides the solution to common API processing tasks such as rate limiting, client authentication, versioning, and logging so that web service developers can focus on the unique features of their services." ~ http://openrepose.org/, 09/03/2013

# Requirements

Supports CentOS 6.5, Ubuntu 14.04, and Ubuntu 12.04.

Requires Repose v6.2.0 or greater due to package differences.

# Usage

Include `recipe[repose]` in your run_list and override attributes as required.

## Filters

There are 2 ways to setup filters:
  1. Include any of the `repose::filter-*` recipes *before* the `repose::default` recipe
  2. Add the filter names to the `node['repose']['filters']` array

Available filters are:
  * [add-header](https://repose.atlassian.net/wiki/display/REPOSE/Add+Header+Filter)
  * [api-validator](https://repose.atlassian.net/wiki/display/REPOSE/API+Validation+filter)
  * [client-auth](https://repose.atlassian.net/wiki/display/REPOSE/Client+Authentication+filter) (deprecated in Repose 8.0.0 - replace with Keystone v2 Filter)
  * [client-authorization](https://repose.atlassian.net/wiki/display/REPOSE/Client+Authorization+filter)
  * [content-type-stripper](https://repose.atlassian.net/wiki/display/REPOSE/Content+Type+Stripper+filter) (system-model only)
  * [CORS](https://repose.atlassian.net/wiki/display/REPOSE/CORS+filter)
  * [DeRP](https://repose.atlassian.net/wiki/display/REPOSE/Delegation+Response+Processor+%28DeRP%29+Filter)
  * [header-identity](https://repose.atlassian.net/wiki/display/REPOSE/Header+Identity+filter)
  * [header-normalization](https://repose.atlassian.net/wiki/display/REPOSE/Header+Normalization+filter)
  * [header-translation](https://repose.atlassian.net/wiki/display/REPOSE/Header+Translation+filter)
  * [ip-identity](https://repose.atlassian.net/wiki/display/REPOSE/IP+Identity+filter) (deprecated in Repose 8.0.0 - replace with IP User Filter)
  * [ip-user](https://repose.atlassian.net/wiki/display/REPOSE/IP+User+filter)
  * [keystone-v2](https://repose.atlassian.net/wiki/display/REPOSE/Keystone+v2+Filter)
  * [merge-header](https://repose.atlassian.net/wiki/display/REPOSE/Merge+Header+filter)
  * [rackspace-auth-user](https://repose.atlassian.net/wiki/display/REPOSE/Rackspace+Auth+User+filter)
  * [rate-limiting](https://repose.atlassian.net/wiki/display/REPOSE/Rate+Limiting+filter)
  * [slf4j-http-logging](https://repose.atlassian.net/wiki/display/REPOSE/SLF4J+HTTP+Logging+filter)
  * [translation](https://repose.atlassian.net/wiki/display/REPOSE/Translation+filter)
  * [uri-identity](https://repose.atlassian.net/wiki/display/REPOSE/URI+Identity+filter)
  * [uri-stripper](https://repose.atlassian.net/wiki/display/REPOSE/URI+Stripper+filter)
  * [url-extractor-to-header](https://repose.atlassian.net/wiki/display/REPOSE/URL+Extractor+to+Header)

Other filters are available in Repose and may be added to this cookbook in a later revision.

## Services

Services work the same way as filters. Just s/filter/service/g.

Available services are:
  * [dist-datastore] (https://repose.atlassian.net/wiki/display/REPOSE/Datastores#Datastores-distributedDistributedDatastore)
  * [http-connection-pool] (https://repose.atlassian.net/wiki/display/REPOSE/HTTP+Connection+Pool+service) (configuration only)
  * [response-messaging] (https://repose.atlassian.net/wiki/display/REPOSE/Response+Messaging+service)

## Nodes

There are also 2 ways to setup nodes in `system-model.cfg.xml`:
  1. Override the `node['repose']['peers']` array as shown below. Note that peers without a `cluster_id` are assumed to belong to all clusters. This is almost certainly not correct if more than 1 cluster is being configured but is implemented to maintain functionality with earlier versions of the cookbook.
  2. Set `node['repose']['peer_search_enabled']` to `true`. For each Chef node found this way peers will be loaded according to the `node['repose']['cluster_ids']`, `node['repose']['node_id']`, `node['fqdn']`, `node['repose']['port']`, and `node['repose']['ssl_port']` attributes. `node_id` and `fqdn` will be consistant for each `cluster_id` in `cluster_ids`. `port` and `ssl_port` will be incremented by 1 for each `cluster_id` in `cluster_ids`.

## Endpoints

Setup endpoint in `system-model.cfg.xml` by overriding the `node['repose']['endpoints']` array as shown below. Note that endpoints without a `cluster_id` are assumed to belong to all clusters. This is almost certainly not correct if more than 1 cluster is being configured but is implemented to maintain functionality with earlier versions of the cookbook.

# Attributes

## General attributes

* `node['repose']['owner']` - The Repose user. `root` on Ubuntu and `repose` on CentOS.
* `node['repose']['group']` - The Repose primary group. `root` on Ubuntu and `repose` on CentOS.
* `node['repose']['install_opts']` - Command line options used for installing pacakges.
* `node['repose']['loglevel']` - The log level for the main Repose log file.
* `node['repose']['cluster_ids']` - An array of cluster IDs.
* `node['repose']['node_id']` - The node ID.
* `node['repose']['port']` - The port to bind to.
* `node['repose']['ssl_port']` - The SSL port to bind to.
* `node['repose']['config_directory']` - Configuration directory location.
* `node['repose']['log_path']` - Log directory location.
* `node['repose']['pid_file']` - PID file location.
* `node['repose']['java_opts']` - Java options specific to Repose. This is distinct from the system environmment variable `JAVA_OPTS`.
* `node['repose']['shutdown_port']` - The shutdown port to bind to.
* `node['repose']['connection_timeout']` - The connection timeout in ms.
* `node['repose']['read_timeout']` The read timeout in ms.
* `node['repose']['deploy_auto_clean']` - Auto-clean the deploy directory if `true`.
* `node['repose']['filter_check_interval']` - The filter check interval in ms.
* `node['repose']['filters']` - An array of filters to load. Loading the `repose::filter-*` recipes will update this if required.
* `node['repose']['services']` - An array of services to load. Loading the `repose::service-*` recipes will update this if required.

## Repository attributes

* `node['repose']['repo']['managed']` - Whether or not to manage the Repose repository. If `false` the recipes will assume the Repose repository is already configured.
* `node['repose']['repo']['baseurl']` - The base URL of the Repose repository.
* `node['repose']['repo']['gpgkey']` - The GPG key used to sign Repose packages.
* `node['repose']['repo']['gpgcheck']` - Whether or not to check package signatures. `rhel` only.
* `node['repose']['repo']['enabled']` - Whether or not the repository is enabled by default. `rhel` only.

These attributes should have reasonable, platform_family specific defaults.

## Peer attributes

* `node['repose']['peer_search_enabled']` - The `repose::default` recipe loads the `repose::load_peers` recipe if `true`.
* `node['repose']['peer_search_query']` - The query to use to search for Repose peers.
* `node['repose']['peers']` - An array of Repose peers.

The Repose peers array defaults to:
```
[
  { 'cluster_id' => 'repose',
    'id' => 'repose_node1',
    'hostname' => 'localhost',
    'port' => 8080,
  }
]
```

## Endpoint attributes

* `node['repose']['endpoints']` - An array of Repose endponts.

The Repose endpoints array defaults to:
```
[
  { 'cluster_id' => 'repose',
    'id' => 'open_repose',
    'protocol' => 'http',
    'hostname' => 'openrepose.org',
    'port' => 80,
    'root_path' => '/',
    'default' => true,
  }
]
```

## dist-datastore attributes

* `node['repose']['dist_datastore']['cluster_id']` - An array of cluster IDs that use this service or `['all']` for all cluster IDs.
* `node['repose']['dist_datastore']['allow_all']` - Allow all hosts to use the dist-datastore if `true`. Be aware that this basically turns off security around the dist-datastore. Use with care.
* `node['repose']['dist_datastore']['allowed_hosts']` - An array of hosts to whitelist for the dist-datastore.
* `node['repose']['dist_datastore']['port']` - The port to use for dist-datastore. Set to `nil` to leave out the `<port-config>` block. Will be incremented by 1 for each `cluster_id` in `node['repose']['cluster_ids']`.

## slf4j-http-logging attributes

* `node['repose']['slf4j_http_logging']['cluster_id']` - An array of cluster IDs that use this filter or `['all']` for all cluster IDs.
* `node['repose']['slf4j_http_logging']['id']` - The log ID.
* `node['repose']['slf4j_http_logging']['format']` - The log format.

## ip-identity attributes

* `node['repose']['ip_identity']['cluster_id']` - An array of cluster IDs that use this filter or `['all']` for all cluster IDs.
* `node['repose']['ip_identity']['quality']` - The default quality.
* `node['repose']['ip_identity']['white_list_quality']` - The whitelist quality.
* `node['repose']['ip_identity']['white_list_ip_addresses']` - An array of whitelisted IP addresses.

## keystone-v2 attributes

* `node['repose']['keystone_v2']['cluster_id']` - An array of cluster IDs that use this filter or `['all']` for all cluster IDs.
* `node['repose']['keystone_v2']['username_admin']` - Administrator username making auth-n requests for clients.
* `node['repose']['keystone_v2']['password_admin']` - Administrator password making auth-n requests for clients.
* `node['repose']['keystone_v2']['uri']` - target URI (i.e., the URI of your Identity Service endpoint) for authentication requests.
* `node['repose']['keystone_v2']['pool']` - Tells the connection pool service to map to the pool with specified id. If default is chosen, the default connection pool configurations in connection pool service is used.
* `node['repose']['keystone_v2']['groups_in_header']` - Set the user's groups in the x-pp-groups header. If the user has no groups, the header will not be set.
* `node['repose']['keystone_v2']['roles_in_header']` - Set the user's roles in the x-roles header.
* `node['repose']['keystone_v2']['catalog_in_header']` - Set the user's service catalog in the x-catalog header.
* `node['repose']['keystone_v2']['delegating_quality']` - If present, the filter will not send a failing response when an invalid state is reached. Instead, it will add the data relating to the failure to a header and forward the request to be handled by a different filter or service.  If not present, the filter will send a failing response when an invalid state is reached.  The quality, a double between 0 and 1, assigned to the delegation header on delegation. This value will be used to order delegation based on priority when multiple delegations are present.
* `node['repose']['keystone_v2']['white_list']` - A list of URI patterns all users can access.
```
white_list = ['$/v1','$/v1/application.wadl$']
```
* `node['repose']['keystone_v2']['cache']` - A container element for all configuration associated with caching.	Can provide timeouts for token, endpoints, and groups and a list of atom feeds
```
cache = {
  'timeout_variability' => 10,
  'token_timeout' => 10,
  'group_timeout' => 10,
  'endpoints_timeout' => 10,
  'atom_feeds' => [
    'feedOne',
    'feedTwo'
  ]
}
```
* `node['repose']['keystone_v2']['tenant_handling']` - A container element for all configuration associated with tenants.
```
tenant_handling = {
  'send_all_tenant_ids' => false,
  'validate_tenant' => {
     'strip_token_tenant_prefixes' => '/test/',
     'url_extraction_regex' => '/v1/servers/.*'
  },
  'send_tenant_id_quality' = {
     'default' => '1.0',
     'roles' => '0.5',
     'uri' => '0.3'
  }
}
```
* `node['repose']['keystone_v2']['require_service_endpoint']` - If this element is included, authorization will be enforced. The use associated with the provided x-auth-token must have an endpoint meeting the criteria defined by the attributes in this element.
```
require_service_endpoint = {
  'public_url' => 'https://somethingorother',
  'region' => 'DFW',
  'name' => 'myservice',
  'type' => 'myservicecatalogentry'
}
```
* `node['repose']['keystone_v2']['pre_authorized_roles']` - If tenant ID validation is enabled, the Keystone v2 filter can be configured to bypass the check for certain requests. To do so, add the <pre-authorized-roles> element to your configuration file as a child of the root element. This element will have <role> children elements which define the roles for which the tenant ID check does not occur. These roles will be compared to the roles returned in a token from the Keystone v2 service, and if there is a match, the project ID check will be skipped.
```
pre_authorized_roles = ['observer','admin']
```

## client-auth attributes

* `node['repose']['client_auth']['cluster_id']` - An array of cluster IDs that use this filter or `['all']` for all cluster IDs.
* `node['repose']['client_auth']['auth_provider']` - The authentication provider to use, either 'RACKSPACE' for Rackspace cloud, or else 'OPENSTACK' for Keystone
* `node['repose']['client_auth']['username_admin']` - Administrator username making auth-n requests for clients.
* `node['repose']['client_auth']['password_admin']` - Administrator password making auth-n requests for clients.
* `node['repose']['client_auth']['auth_uri']` - URI for the authentication service.
* `node['repose']['client_auth']['tenanted']` - Whether the tenant_id will be parsed from the request uri.  If this is set to true then mapping_regex must be provided.
* `node['repose']['client_auth']['mapping_regex']` - Optional regex to extract the user/tenant/account information from URI.
    * For example this regex `.*/v1/([-|\w]+)/?.*` will provide this action: '/v1/853473/action => x-pp-user=853473'. Used with the 'CLOUD' Rackspace mapping_type below.
    * For example this regex `.*.com/service/([-|\w]+)/?.*` will provide this action: '/service/853473/action => x-pp-user=853473'. Used with the 'MOSSO' Rackspace mapping_type below.
* `node['repose']['client_auth']['mapping_type']` - (Rackspace only, required if 'mapping_regex' is provided) One of 'CLOUD' or 'MOSSO'.
* `node['repose']['client_auth']['tenant_id']` - (Keystone only, optional) Tenant-ID with admin role for the admin user above.
* `node['repose']['client_auth']['delegable']` - Whether auth is delegable.

* `node['repose']['client_auth']['request_groups']` - Whether repose requests the user's list of groups from the identity service.
* `node['repose']['client_auth']['token_cache_timeout']` - Timeout for token cache.
* `node['repose']['client_auth']['group_cache_timeout']` - Timeout for group cache.
* `node['repose']['client_auth']['endpoints_in_header']` - Enable or disable the listing of service endpoints in the header.

* `node['repose']['client_auth']['white_list']` - You can configure Repose authentication to allow the processing of requests that do not require authentication. Default to `false`.
* `node['repose']['client_auth']['uri_regex']` - The whitelist contains a list of regular expressions Repose will attempt to match against the full request URI. If the URI matches a regular expression on the white list, the request is passed to the origin service. Otherwise, authentication is performed against the request.

## rate-limit attributes

* `node['repose']['rate_limiting']['cluster_id']` - An array of cluster IDs that use this filter or `['all']` for all cluster IDs.
* `node['repose']['rate_limiting']['uri_regex']` - A regular expression (regex) for the URI at which the user can query their limits.
* `node['repose']['rate_limiting']['include_absolute_limits']` - Enables or disables integration with absolute limits.
* `node['repose']['rate_limiting']['global_limits']` - An array of global limits.  Defaults to empty thus disabled.
* `node['repose']['rate_limiting']['limit_groups']` - An array of limit groups.

The limit groups array defaults to:
```
[
  { 'id' => 'limited',
    'groups' => 'limited',
    'default' => true,
    'limits' => [
      { 'id' => 'all',
        'uri' => '*',
        'uri_regex' => '/.*',
        'http_methods' => 'POST PUT GET DELETE',
        'unit' => 'MINUTE',
        'value' => 10
      }
    ]
  },
  { 'id' => 'unlimited',
    'groups' => 'unlimited',
    'default' => false,
    'limits' => []
  }
]
```

Global limit groups are similarly organized:
```
[
  { 'id' => 'create-server',
    'uri' => '/server/create',
    'uri-regex' => '/server/create/?',
    'method' => 'POST',
    'value' => 10,
    'unit' => 'SECOND'
  }
]
```

## header-translation attributes

* `node['repose']['header_translation']['headers']` - header list of original headers (the ones passed in the request) and the new headers (the ones that will have duplicate values of original headers).  new_name can be a list of headers white-space separated. You can also set an optional remove_attribute key if you'd like the original header to be removed.  to set the array.

The default headers are:
```
[
  { 'original_name' => 'Content-Type',
    'new_name' => 'rax-content-type'
  },
  { 'original_name' => 'Content-Length',
    'new_name' => 'rax-content-length not-rax-content-length something-else',
    'remove_original' => true
  }
]
```

## header-identity attributes

* `node['repose']['header_identity']['cluster_id']` - An array of cluster IDs that use this filter or `['all']` for all cluster IDs.
* `node['repose']['header_identity']['headers']` - Lists the headers to identify the incoming user.

The default headers are:
```
[
  { 'id' => 'X-Auth-Token',
    'quality' => 0.95
  },
  { 'id' => 'X-Forwarded-For',
    'quality' => 0.5
  }
]
```

## header-normalization attributes

* `node['repose']['header_normalization']['whitelist']` - a hash of whitelist blocks, each encapsulating an array of headers to be whitelisted by Repose for delivery to the origin service or next filter in the chain. A unique 'id' key identifies each whitelist entry, with optional 'uri_regex' and 'http_methods' keys to induce further constraints.
* `node['repose']['header_normalization']['blacklist']` - a hash of blacklist blocks, each encapsulating an array of headers to be blacklisted by Repose for non-delivery to the origin service or next filter in the chain. A unique 'id' key identifies each blacklist entry, with optional 'uri_regex' and 'http_methods' keys to induce further constraints.

The default whitelist is:
```
[
  { 'id' => 'credentials',
    'uri_regex' => '/servers/(.*)',
    'headers'   => ['X-Auth-Key','X-Auth-User']
  },
  { 'id' => 'modification',
    'uri_regex' => '/resource/(.*)',
    'http_methods' => 'PUT POST',
    'headers'   => ['X-Modify']
  }
]
```

The default blacklist is:
```
[
  { 'id' => 'rate-limit-headers',
    'headers'   => ['X-PP-User','X-PP-Groups']
  }
]
```

## merge-header attributes

* `node['repose']['merge_header']['cluster_id']` - An array of cluster IDs that use this filter or `['all']` for all cluster IDs.
* `node['repose']['merge_header']['response_headers']` - An array of response headers to merge into one
* `node['repose']['merge_header']['request_headers']` - An array of response headers to merge into one

The default response_headers and request_headers are:
```
[]
```
which means no headers to merge and this filter becomes a no-op. An example to override it:
```
node['repose']['merge_header']['response_headers'] = [
    'HEADER1', 'HEADER2'
]
node['repose']['merge_header']['request_headers'] = [
    'HEADER1', 'HEADER2'
]
```

## content-type-stripper attributes

* `node['repose']['content_type_stripper']['cluster_id']` - An array of cluster IDs that use this filter or `['all']` for all cluster IDs.

## CORS attributes

* `node['repose']['cors']['cluster_id']` - An array of cluster IDs that use this filter or `['all']` for all cluster IDs.
* `node['repose']['cors']['allowed_origins']` - An array of origin objects
* `node['repose']['cors']['allowed_methods']` - An array of methods that are allowed to access origin service
* `node['repose']['cors']['resources']` - An array of resource specific CORS that are allowed to be accessed. Each resource consists of path and an array of allowed_methods 

The default allowed_origins is:
```
[
  { 'is_regex' => true,
    'value'    => '.*'
  }
]
```
which means all origins are allowed

The default allowed_methods is:
```
[]
```
which means all methods are allowed

The default resources is:
```
[]
```
which means all resources are allowed to be accessed. An example configuration that overrides the default resources:
```
node['repose']['cors']['resources'] = [
  { 'path'            => '/resource1/.*',
    'allowed_methods' => [
      'GET'
     ]
  }
]
```
which means only GET methods are allowed for CORS requests to "/resource1/.*"


## http-connection-pool attributes


CHANGELOG.txt
v3.3.9
------
* add support for merge-header and CORS filters

v3.3.8
------
* update default mode on files which have credentials

v3.3.7
------
* fix allowed_hosts sorting bug which broke dist-datastore

v3.3.6
------

v3.3.5
------
* fix bug where overLimit-429-responsecode attribute had incorrect casing (rate-limiting.cfg.xml template)
* change default logging to WARN instead of DEBUG (log4j2.xml.erb)
* change default log rotation to 512MB and 4 files (log4j2.xml.erb)

v3.3.4
------
* update log4j2.xml template to use numbers in log rotation instead of date
* pin default version to 7.3.6.0 to avoid JRE 8 issues (until tested later)
* filter template updates - api-validator, rate-limiting, keystone-v2, slf4j-http-logging

v3.3.3
------
* add optional content-body-read-limit, client-request-logging and proxy-thread-pool options to the container.cfg.xml template, set values to nil
* added 10 additional options for the api-validator filter
* fix minor bug with manually declared peers and system-model template

v3.3.2
------
* minor code cleanup (removing unneeded include_recipe statements in filter and service recipes)
* set deploy_auto_clean to true
* remove minitest artifacts

v3.3.1
------
* Finish adding basic unit tests for existing filter and service recipes
* Rename service-connection-pool to service-http-connection-pool to match current naming schemes
* Added CHANGELOG and TODO files 
* set hosts, nodes to sort in system-config and dist-datastores


attributes/default.rb
case node['platform_family']
when 'rhel'
  default['repose']['owner'] = 'repose'
  default['repose']['group'] = 'repose'
  default['repose']['repo']['baseurl'] = 'http://repo.openrepose.org/rhel'
  default['repose']['repo']['gpgkey'] = 'http://repo.openrepose.org/rhel/RPM_GPG_KEY-openrepose'
  default['repose']['repo']['gpgcheck'] = false # the openrepose repo doesn't sign packages
  default['repose']['repo']['enabled'] = true
  default['repose']['repo']['managed'] = true
  default['repose']['install_opts'] = ''
when 'debian'
  default['repose']['owner'] = 'root'
  default['repose']['group'] = 'root'
  default['repose']['repo']['baseurl'] = 'http://repo.openrepose.org/debian'
  default['repose']['repo']['gpgkey'] = 'http://repo.openrepose.org/debian/pubkey.gpg'
  default['repose']['repo']['managed'] = true
  default['repose']['install_opts'] = '-o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" --force-yes'
end

default['repose']['version'] = '7.3.6.0'
default['repose']['loglevel'] = 'WARN'
default['repose']['cluster_ids'] = ['repose']
default['repose']['rewrite_host_header'] = true
default['repose']['node_id'] = 'repose_node1'
default['repose']['port'] = 8080
default['repose']['ssl_port'] = 8443
default['repose']['shutdown_port'] = 8188
# container.cfg.xml settings
default['repose']['content_body_read_limit'] = nil
default['repose']['connection_timeout'] = 30000
default['repose']['read_timeout'] = 30000
default['repose']['client_request_logging'] = nil
default['repose']['proxy_thread_pool'] = nil
default['repose']['deploy_auto_clean'] = true
default['repose']['filter_check_interval'] = 60000
#
default['repose']['config_directory'] = '/etc/repose'
default['repose']['log_path'] = '/var/log/repose'
default['repose']['pid_file'] = '/var/run/repose-valve.pid'
default['repose']['user'] = 'repose'
default['repose']['java_opts'] = ''
default['repose']['loggers'] = []
default['repose']['appenders'] = []

default['repose']['peer_search_enabled'] = false
default['repose']['peer_search_query'] = "chef_environment:#{node.chef_environment} AND repose_cluster_ids:*"

default['repose']['peers'] = [
  { 'cluster_id' => 'repose',
    'id' => 'repose_node1',
    'hostname' => 'localhost',
    'port' => 8080 }
]

default['repose']['filters'] = []
default['repose']['services'] = []

default['repose']['endpoints'] = [
  { 'cluster_id' => 'repose',
    'id' => 'open_repose',
    'protocol' => 'http',
    'hostname' => 'openrepose.org',
    'port' => 80,
    'root_path' => '/',
    'default' => true }
]

default['repose']['add_header']['cluster_id'] = ['all']
default['repose']['add_header']['request_headers'] = []
default['repose']['add_header']['response_headers'] = []

default['repose']['url_extractor_to_header']['cluster_id'] = ['all']
default['repose']['url_extractor_to_header']['extractor_headers'] = []

default['repose']['dist_datastore']['cluster_id'] = ['all']
default['repose']['dist_datastore']['allow_all'] = false
default['repose']['dist_datastore']['allowed_hosts'] = ['127.0.0.1']
default['repose']['dist_datastore']['port'] = 8081

default['repose']['response_messaging']['cluster_id'] = ['all']
default['repose']['response_messaging']['status_codes'] = []

default['repose']['slf4j_http_logging']['cluster_id'] = ['all']
default['repose']['slf4j_http_logging']['id'] = 'http'
default['repose']['slf4j_http_logging']['format'] = 'Remote IP=%a Local IP=%A Response Size(bytes)=%b Remote Host=%h Request Method=%m Server Port=%p Query String=%q Time Request Received=%t Status=%s Remote User=%u Rate Limit Group: %{X-PP-Groups}i URL Path Requested=%U X-Forwarded-For=%{X-Forwarded-For}i X-REAL-IP=%{X-Real-IP}i'

# default for merge_header is empty (no-op)
default['repose']['merge_header']['cluster_id'] = ['all']

# default for cors is to allow all methods, all origins
default['repose']['cors']['cluster_id'] = ['all']
default['repose']['cors']['allowed_origins'] = [
  { 'is_regex' => true,
    'value'    => '.*' }
]

default['repose']['header_translation']['cluster_id'] = ['all']
default['repose']['header_translation']['headers'] = [
  { 'original_name' => 'Content-Type',
    'new_name' => 'rax-content-type' },
  { 'original_name' => 'Content-Length',
    'new_name' => 'rax-content-length not-rax-content-length something-else',
    'remove_original' => true }
]

default['repose']['derp']['cluster_id'] = ['all']

default['repose']['content_type_stripper']['cluster_id'] = ['all']

default['repose']['header_normalization']['cluster_id'] = ['all']
default['repose']['header_normalization']['whitelist'] = [
  { 'id' => 'credentials',
    'uri_regex' => '/servers/(.*)',
    'headers'   => ['X-Auth-Key', 'X-Auth-User'] },
  { 'id' => 'modification',
    'uri_regex' => '/resource/(.*)',
    'http_methods' => 'PUT POST',
    'headers'   => ['X-Modify'] }
]
default['repose']['header_normalization']['blacklist'] = [
  { 'id' => 'rate-limit-headers',
    'headers'   => ['X-PP-User', 'X-PP-Groups'] }
]

default['repose']['header_identity']['cluster_id'] = ['all']
default['repose']['header_identity']['headers'] = [
  { 'id' => 'X-Auth-Token',
    'quality' => 0.95 },
  { 'id' => 'X-Forwarded-For',
    'quality' => 0.5 }
]

default['repose']['ip_identity']['cluster_id'] = ['all']
default['repose']['ip_identity']['quality'] = 0.2
default['repose']['ip_identity']['white_list_quality'] = 1.0
default['repose']['ip_identity']['white_list_ip_addresses'] = ['127.0.0.1']

default['repose']['keystone_v2']['cluster_id'] = ['all']
default['repose']['keystone_v2']['username_admin'] = 'admin'
default['repose']['keystone_v2']['password_admin'] = 'password'
default['repose']['keystone_v2']['pool'] = nil
default['repose']['keystone_v2']['groups_in_header'] = nil
default['repose']['keystone_v2']['roles_in_header'] = nil
default['repose']['keystone_v2']['catalog_in_header'] = nil
default['repose']['keystone_v2']['uri'] = 'https://identity.api.rackspacecloud.com/v2.0'
default['repose']['keystone_v2']['delegating_quality'] = nil
default['repose']['keystone_v2']['white_list'] = nil
default['repose']['keystone_v2']['cache'] = nil
default['repose']['keystone_v2']['tenant_handling'] = nil
default['repose']['keystone_v2']['require_service_endpoint'] = nil
default['repose']['keystone_v2']['pre_authorized_roles'] = nil

default['repose']['client_auth']['cluster_id'] = ['all']
default['repose']['client_auth']['auth_provider'] = 'RACKSPACE'
default['repose']['client_auth']['username_admin'] = 'admin'
default['repose']['client_auth']['password_admin'] = 'password'
default['repose']['client_auth']['tenant_id'] = 'tenant-id'
default['repose']['client_auth']['auth_uri'] = 'https://auth.api.rackspacecloud.com/v1.1/auth'
default['repose']['client_auth']['mapping_regex'] = ['.*/v1/([-|\w]+)/?.*']
default['repose']['client_auth']['mapping_type'] = 'CLOUD'
default['repose']['client_auth']['delegable'] = false
default['repose']['client_auth']['tenanted'] = true
default['repose']['client_auth']['request_groups'] = true
default['repose']['client_auth']['token_cache_timeout'] = 600000
default['repose']['client_auth']['group_cache_timeout'] = 600000
default['repose']['client_auth']['service_admin_roles'] = []
default['repose']['client_auth']['ignore_tenant_roles'] = []
default['repose']['client_auth']['endpoints_in_header'] = false
default['repose']['client_auth']['white_list'] = false
default['repose']['client_auth']['uri_regex'] = nil

default['repose']['client_authorization']['cluster_id'] = 'all'
default['repose']['client_authorization']['username_admin'] = 'admin'
default['repose']['client_authorization']['password_admin'] = 'password'
default['repose']['client_authorization']['auth_uri'] = 'https://auth.api.rackspacecloud.com/v1.1/auth'
default['repose']['client_authorization']['tenant_id_admin'] = nil
default['repose']['client_authorization']['endpoint_list_ttl'] = 300
default['repose']['client_authorization']['connection_pool_id'] = nil
default['repose']['client_authorization']['service_endpoint'] = 'https://exampleservice.api.rackspacecloud.com/v1.0'
default['repose']['client_authorization']['service_region'] = nil
default['repose']['client_authorization']['service_name'] = nil
default['repose']['client_authorization']['service_type'] = nil
default['repose']['client_authorization']['ignore_tenant_roles'] = []
default['repose']['client_authorization']['roles'] = []
default['repose']['client_authorization']['delegating_quality'] = nil

default['repose']['rate_limiting']['cluster_id'] = ['all']
default['repose']['rate_limiting']['datastore'] = nil
default['repose']['rate_limiting']['datastore_warn_limit'] = nil
default['repose']['rate_limiting']['overlimit_429_responsecode'] = nil
default['repose']['rate_limiting']['use_capture_groups'] = nil
default['repose']['rate_limiting']['uri_regex'] = '/limits'
default['repose']['rate_limiting']['global_limits'] = []
default['repose']['rate_limiting']['include_absolute_limits'] = false
default['repose']['rate_limiting']['limit_groups'] = [
  { 'id' => 'limited',
    'groups' => 'limited',
    'default' => true,
    'limits' => [
      { 'id' => 'all',
        'uri' => '*',
        'uri_regex' => '/.*',
        'http_methods' => 'POST PUT GET DELETE',
        'unit' => 'MINUTE',
        'value' => 10 }
    ] },
  { 'id' => 'unlimited',
    'groups' => 'unlimited',
    'default' => false,
    'limits' => [] }
]

default['repose']['http_connection_pool']['cluster_id'] = ['all']
default['repose']['http_connection_pool']['chunked_encoding'] = true
default['repose']['http_connection_pool']['max_total'] = 400
default['repose']['http_connection_pool']['max_per_route'] = 200
default['repose']['http_connection_pool']['socket_timeout'] = 30000
default['repose']['http_connection_pool']['connection_timeout'] = 30000
default['repose']['http_connection_pool']['socket_buffer_size'] = 8192
default['repose']['http_connection_pool']['connection_max_line_length'] = 8192
default['repose']['http_connection_pool']['connection_max_header_count'] = 100
default['repose']['http_connection_pool']['connection_max_status_line_garbage'] = 100
default['repose']['http_connection_pool']['tcp_nodelay'] = true
default['repose']['http_connection_pool']['keepalive_timeout'] = 0

default['repose']['translation']['cluster_id'] = ['all']
default['repose']['translation']['allow_doctype_decl'] = false
default['repose']['translation']['request_translations'] = []
default['repose']['translation']['response_translations'] = []

default['repose']['rackspace_auth_user']['cluster_id'] = ['all']
default['repose']['rackspace_auth_user']['v1_1']['enabled'] = true
default['repose']['rackspace_auth_user']['v1_1']['read_limit'] = 4096
default['repose']['rackspace_auth_user']['v1_1']['group'] = 'Pre_Auth'
default['repose']['rackspace_auth_user']['v1_1']['quality'] = 0.6
default['repose']['rackspace_auth_user']['v2_0']['enabled'] = true
default['repose']['rackspace_auth_user']['v2_0']['read_limit'] = 4096
default['repose']['rackspace_auth_user']['v2_0']['group'] = 'Pre_Auth'
default['repose']['rackspace_auth_user']['v2_0']['quality'] = 0.6

default['repose']['uri_identity']['cluster_id'] = ['all']
default['repose']['uri_identity']['mappings'] = [
  '.*/servers/(\w*)/.*',
  '.*/servers/(\w*)/.*'
]
default['repose']['uri_identity']['group'] = 'User_Standard'
default['repose']['uri_identity']['quality'] = 0.5

default['repose']['uri_stripper']['cluster_id'] = ['all']
default['repose']['uri_stripper']['rewrite_location'] = true
default['repose']['uri_stripper']['token_index'] = 1

default['repose']['add_header']['cluster_id'] = ['all']
default['repose']['add_header']['requests'] = []
default['repose']['add_header']['responses'] = []

default['repose']['api_validator']['cluster_id'] = ['all']
default['repose']['api_validator']['multi_role_match'] = nil
default['repose']['api_validator']['wadl'] = nil
default['repose']['api_validator']['dot_output'] = '/tmp/default.out'
default['repose']['api_validator']['enable_rax_roles'] = nil
default['repose']['api_validator']['enable_api_coverage'] = nil
default['repose']['api_validator']['check_well_formed'] = nil
default['repose']['api_validator']['check_grammars'] = nil
default['repose']['api_validator']['check_elements'] = nil
default['repose']['api_validator']['check_plain_params'] = nil
default['repose']['api_validator']['do_xsd_grammar_transform'] = nil
default['repose']['api_validator']['enable_pre_process_extension'] = nil
default['repose']['api_validator']['remove_dups'] = nil
default['repose']['api_validator']['xpath_version'] = nil
default['repose']['api_validator']['xsl_engine'] = nil
default['repose']['api_validator']['xsd_engine'] = nil
default['repose']['api_validator']['mask_rax_roles_403'] = nil
default['repose']['api_validator']['join_xpath_checks'] = nil
default['repose']['api_validator']['enable_ignore_xsd_extension'] = nil
default['repose']['api_validator']['validator_name'] = nil
default['repose']['api_validator']['validate_checker'] = nil
default['repose']['api_validator']['check_headers'] = nil

default['repose']['ip_user']['cluster_id'] = ['all']
default['repose']['ip_user']['user_header_name'] = 'X-PP-User'
default['repose']['ip_user']['user_header_quality'] = '0.4'
default['repose']['ip_user']['group_header_name'] = 'X-PP-Groups'
default['repose']['ip_user']['group_header_quality'] = '0.4'
default['repose']['ip_user']['groups'] = [
  { 'group_name' => 'match-all',
    'cidrips' => ['0.0.0.0/0'] },
  { 'group_name' => 'rfc1918',
    'cidrips' => ['192.168.0.0/16', '172.16.0.0/12', '10.0.0.0/8'] }
]
